openedx/frontend-app-authoring
# feat: shareable URLs for library components and searches [FC-0076]

## Description

This PR adds new routes and URL parameters to use when viewing and performing searches on library components. These changes allow these pages to be bookmarked or shared by copy/pasting the browser's current URL.

No changes were made to the UI.

Use cases covered:

- [x] As an author working with content libraries, I want to easily share any component in a library with other people on my team, by copying the URL from my browser and sending it to them.
- [x] As an author working with content libraries, I want to easily share any search results with other people on my team, by copying the URL from my browser and sending it to them.
- [x] As an author working with content libraries, I want to bookmark a search in my browser and return to it at any time, with the same filters and keywords applied.
- [x] As an author of a content library with public read access, I want to easily share any component in a library with any authors on the same Open edX instance, by copying the URL from my browser and sending it to them.
- [x] As an author of a content library, I want to easily share a library's "Manage Team" page with other people on my team by copying the URL from my browser and sending it to them.
- [x] As an author working with content libraries, I want to easily share any selected sidebar tab with other people on my team, by copying the URL from my browser and sending it to them.

## Supporting information

See https://github.com/openedx/frontend-app-authoring/issues/1499
Private-ref: [FAL-3984](https://tasks.opencraft.com/browse/FAL-3984)

## Testing instructions

Code review: Recommend going commit-by-commit, since this is a pretty big change.

This change involved some route additions and changes, and so to manually test them, please refresh your browser page after each step, and ensure you see exactly the same page -- including selected tabs, sidebar, and sidebar tabs. Please also check that using the back/forth buttons navigate smoothly between pages as expected, but that changing the library search parameters doesn't append to the browser history.

1. Navigate to `/libraries` and create or select an existing library from the list.
   "All Content" should still be the default tab you see.
2. Create a few components, collections, and add some components to some collections.
   Note that the component picker still works as expected.
3. Navigate back to "All Content".
4. Select a component from the list.
   Note that the URL changes to `/library/:libraryId/component/:componentId`, and the component is visible in the sidebar.
5. Add or select a different component from the list.
   Note that the URL changes to reflect the new componentId.
6. Add or select a collection from the list.
   Note that the URL changes to `/library/:libraryId/:collectionId`, and the collection is visible in the sidebar.
7. Add or select a new collection from the list.
   Note that the URL changes to reflect the new collectionId.
8. Change your sidebar tab to something other than the initial Preview, then select a different component from the list.
   Note that the sidebar tab is preserved (where possible) when navigating between components or collections.
   Clicking Library Info should reset the sidebar back to the Library.
9. Change to the Components and Collections tabs.
   Note that you remain in the selected tab when switching between selected components/collections.
10. Select a Collection, and either click the Open button in its sidebar or click the Collection card again to open it.
   Note that the URL changes to `/library/:libraryId/collection/:collectionId`, and the collection is visible in the sidebar.
11. Select a component within the collection.
   Note that the URL changes to `/library/:libraryId/collection/:collectionId/:componentId`, and the component is visible in the sidebar.
12. Change your sidebar tab to something other than the initial Preview, then select a different component from the list.
   Note that the sidebar tab is preserved (where possible) when navigating between components.
   Clicking Collection Info should reset the sidebar back to the Collection.
13. Use the search bar keywords to search for components from any tab or page.
   Note that your keywords are stored in a `q` query string parameter, and preserved on page refresh.
14. Repeat for the other search filters and sort options.
   Note that these are store in the query string, and cleared when de-selecting or clicking Clear Filters.
15. Navigate back to the Library sidebar, and click Manage Access.
   Note that `?sa=manage-team` is the query string, and refreshing the page re-opens the Library Team modal.
16. Using the Component card menu, select Add to Collection.
   Note that `?sa=jump-to-add-collections` is in the query string, and this is preserved when you navigate between components.
   Clicking "Confirm" or "Cancel" removes this parameter and closes the collection manager.
17. Check that the Problem Bank and Library Content pickers still work as expected.

## Author Notes & Concerns

1. When https://github.com/openedx/frontend-app-authoring/pull/1570 merges, we'll need to apply the same "URL search params" treatment to the "Publish Status" filter that we've done with the other filters.
2. I chose not to store whether the sidebar was open or closed in the query string, because the sidebar reopens itself almost everywhere in library authoring, so it seemed unnecessary.
3. I left some `TODO`s in the SearchManager code about sanitizing parameters, because sanitising user input is good practice. But I'm not sure how best to sanitise things like search keywords?
4. I ran into a really tricky bug when running `clearFilters` on multiple search filters that are now using the  `useStateWithUrlSearchParam` hook  -- more details in [this comment](https://github.com/openedx/frontend-app-authoring/pull/1575/files#diff-a2d969770fbca6d31b2bcb062b5d406aa54f9ddebef57996faacc6f82e746e42R111-R127). Advice or alternatives welcome!
5. I tried to add tests to validate the navigation/routes used in this PR, but path changes aren't visible with the `<MemoryRouter>` used with `testUtils`, and none of the other available `<Router>` classes would let me set the base `libraryId` path. Advice welcome :)
